sdk/lib/rtl/rangelist.c

```diff
@@ -13,16 +13,478 @@
 #define NDEBUG
 #include <debug.h>
 
+/* GLOBALS ******************************************************************/
+
+//extern PAGED_LOOKASIDE_LIST RtlpRangeListEntryLookasideList;
+
 /* TYPES ********************************************************************/
 
-typedef struct _RTL_RANGE_ENTRY
-{
-    LIST_ENTRY Entry;
-    RTL_RANGE Range;
-} RTL_RANGE_ENTRY, *PRTL_RANGE_ENTRY;
+/* RTLP_RANGE_LIST_ENTRY == RTL_RANGE + ListEntry */
+
+/*  FIXME ? ntddk.h */
+/* Flags for public functions: RtlFindRange(), RtlIsRangeAvailable() */
+#define RTL_RANGE_LIST_SHARED_OK           1
+#define RTL_RANGE_LIST_NULL_CONFLICT_OK    2
+/* ? ntddk.h */
+
+/* RTLP_RANGE_LIST_ENTRY.PrivateFlags */
+#define RTLP_ENTRY_IS_MERGED  1
+
 
 /* FUNCTIONS ***************************************************************/
 
+static
+PRTLP_RANGE_LIST_ENTRY
+FASTCALL
+RtlpEntryFromLink(PVOID Link)
```

Comment:
I think in this case it's more clear to use `CONTAINING_RECORD` directly, without any wrapper functions

Reply:
that, or making the function `inline`

Reply:
No, the point is it makes code understanding more difficult (because wraps a single simple operation in a function)

Reply:
I think the opposite.
`For the usual case, it will be something like this:

    for (MergedRtlEntry = RtlpEntryFromLink(RtlEntry->Merged.ListHead.Flink);
         &MergedRtlEntry->ListEntry != &RtlEntry->Merged.ListHead;
         MergedRtlEntry = RtlpEntryFromLink(MergedRtlEntry->ListEntry.Flink))
         
    for (MergedRtlEntry = CONTAINING_RECORD(RtlEntry->Merged.ListHead.Flink, RTLP_RANGE_LIST_ENTRY, ListEntry);
         &MergedRtlEntry->ListEntry != &RtlEntry->Merged.ListHead;
         MergedRtlEntry = CONTAINING_RECORD(MergedRtlEntry->ListEntry.Flink, RTLP_RANGE_LIST_ENTRY, ListEntry))
`
`For a safe case - like this:

    for (CurrentRtlEntry = RtlpEntryFromLink(RtlEntry->ListEntry.Flink),  NextRtlEntry = RtlpEntryFromLink(CurrentRtlEntry->ListEntry.Flink);
         &CurrentRtlEntry->ListEntry != ListHead;
         CurrentRtlEntry = NextRtlEntry, NextRtlEntry = RtlpEntryFromLink(CurrentRtlEntry->ListEntry.Flink))

    for (CurrentRtlEntry = CONTAINING_RECORD(RtlEntry->ListEntry.Flink, RTLP_RANGE_LIST_ENTRY, ListEntry),
         NextRtlEntry = CONTAINING_RECORD(CurrentRtlEntry->ListEntry.Flink, RTLP_RANGE_LIST_ENTRY, ListEntry);
         &CurrentRtlEntry->ListEntry != ListHead;
         CurrentRtlEntry = NextRtlEntry,
         NextRtlEntry = CONTAINING_RECORD(CurrentRtlEntry->ListEntry.Flink, RTLP_RANGE_LIST_ENTRY, ListEntry))
`
But if you insist, then I will make a replacement.